core/src/main/java/org/infinispan/stream/impl/tx/TxClusterStreamManager.java

```diff
@@ -90,19 +98,60 @@ public void forgetOperation(Object id) {
       return manager.receiveResponse(id, origin, complete, segments, response);
    }
 
+   @Override
+   public <E> RemoteIteratorPublisher<E> remoteIterationPublisher(boolean parallelStream,
+         Supplier<Map.Entry<Address, IntSet>> segments, Set<K> keysToInclude, IntFunction<Set<K>> keysToExclude,
+         boolean includeLoader, Iterable<IntermediateOperation> intermediateOperations) {
+
+      if (ctx.getLookedUpEntries().isEmpty()) {
+         return manager.remoteIterationPublisher(parallelStream, segments, keysToInclude, keysToExclude, includeLoader,
+               intermediateOperations);
+      } else {
+         Set<K>[] contextSet = generateContextSet(ctx);
+         if (keysToExclude == null) {
+            return manager.remoteIterationPublisher(parallelStream, segments, keysToInclude, i -> contextSet[i], includeLoader,
+                  intermediateOperations);
+         } else {
+            return manager.remoteIterationPublisher(parallelStream, segments, keysToInclude, i -> {
+               Set<K> set = contextSet[i];
+               if (set != null) {
+                  set.addAll(keysToExclude.apply(i));
+                  return set;
+               } else {
+                  return keysToExclude.apply(i);
+               }
+            }, includeLoader, intermediateOperations);
+         }
+      }
+   }
+
+   Set<K>[] generateContextSet(LocalTxInvocationContext ctx) {
+      Set<K>[] set = new Set[maxSegments];
+      ctx.getLookedUpEntries().forEach((k, v) -> {
+         int segment = intFunction.applyAsInt(k);
+         Set<K> innerSet = set[segment];
+         if (innerSet == null) {
+            innerSet = new HashSet<>();
+            set[segment] = innerSet;
+         }
+         innerSet.add((K) k);
+      });
+      return set;
+   }
+
    private static class TxExcludedKeys<K> extends AbstractDelegatingMap<Integer, Set<K>> {
       private final Map<Integer, Set<K>> map;
       private final Map<Integer, Set<K>> ctxMap;
 
-      private TxExcludedKeys(Map<Integer, Set<K>> map, LocalTxInvocationContext ctx, ConsistentHash hash) {
+      private TxExcludedKeys(Map<Integer, Set<K>> map, LocalTxInvocationContext ctx, ToIntFunction<Object> intFunction) {
          this.map = map;
-         this.ctxMap = contextToMap(ctx, hash);
+         this.ctxMap = contextToMap(ctx, intFunction);
```

Comment:
Not sure I understand the question. The `convertToMap` eagerly converts all the entries into a map keyed by the segment, so that we can very easily find which entries map to which. This is important as the context could be checked multiple times and we don't want to compute it every time.

Reply:
The `ctxMap` is always generated in the constructor of `TxExcludedKeys`, so why can't we just have the `contextToMap` function in the scope of `TxClusterStreamManager` and pass the calculated `ctxMap` to the constructor of `TxExcludedKeys`. That way we don't have to pass both the `ctx` and `intFunction` instances to the constructor of `TxExcludedKeys` (or to `contextToMap` as both are class variables).

Furthermore, and I'm not sure if this is the case, but if `ctx.getLookedUpEntries()` will always return the same entries for a given ctx, then we can just compute `ctxMap` once.

Reply:
In this case we have to calculate this map per invocation. The ctx can and will change between calls. It can even change between the constructor being called and invoking the remote stream operation.

Reply:
TxClusterStreamManager constructor is called when user invokes `stream` or `parallelStream` in a transactional context. There is nothing preventing them from then changing the context before invoking a terminal operation on that stream (which finally calls `remoteStream*` methods).